hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestReplicationEditsDroppedWithDeletedTableCFs.java

```diff
@@ -0,0 +1,250 @@
+    // add peer
+    ReplicationPeerConfig rpc = ReplicationPeerConfig.newBuilder()
+        .setClusterKey(utility2.getClusterKey())
+        .setReplicateAllUserTables(true).build();
+    admin1.addReplicationPeer(PEER_ID, rpc);
+    // create table
+    createTable();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    // Remove peer
+    admin1.removeReplicationPeer(PEER_ID);
+    Thread.sleep(SLEEP_TIME);
+//    utility1.getMiniHBaseCluster().getMaster().getReplicationZKNodeCleanerChore().choreForTesting();
+    Thread.sleep(SLEEP_TIME);
+    // Drop table
+    admin1.disableTable(TABLE);
+    admin1.deleteTable(TABLE);
+    admin2.disableTable(TABLE);
+    admin2.deleteTable(TABLE);
+  }
+
+  private void createTable() throws Exception {
+    TableDescriptor desc = createTableDescriptor(NORMAL_CF, DROPPED_CF);
+    admin1.createTable(desc);
+    admin2.createTable(desc);
+    utility1.waitUntilAllRegionsAssigned(desc.getTableName());
+    utility2.waitUntilAllRegionsAssigned(desc.getTableName());
+  }
+
+  @Test
+  public void testEditsDroppedWithDeleteCF() throws Exception {
+    admin1.disableReplicationPeer(PEER_ID);
+
+    try (Table table = utility1.getConnection().getTable(TABLE)) {
+      Put put = new Put(ROW);
+      put.addColumn(DROPPED_CF, QUALIFIER, VALUE);
+      table.put(put);
+    }
+
+    deleteCf(admin1);
+
+    admin1.enableReplicationPeer(PEER_ID);
+
+    verifyReplicationProceeded();
+  }
+
+  @Test
+  public void testEditsBehindDeleteCFTiming() throws Exception {
+    admin1.disableReplicationPeer(PEER_ID);
+
+    try (Table table = utility1.getConnection().getTable(TABLE)) {
+      Put put = new Put(ROW);
+      put.addColumn(DROPPED_CF, QUALIFIER, VALUE);
+      table.put(put);
+    }
+
+    // Only delete cf from peer cluster
+    deleteCf(admin2);
+
+    admin1.enableReplicationPeer(PEER_ID);
+
+    // the source table's cf still exists, replication should be stalled
+    verifyReplicationStuck();
+    deleteCf(admin1);
+    // now the source table's cf is gone, replication should proceed, the
+    // offending edits be dropped
+    verifyReplicationProceeded();
+  }
+
+  private void verifyReplicationProceeded() throws Exception {
+    try (Table table = utility1.getConnection().getTable(TABLE)) {
+      Put put = new Put(ROW);
+      put.addColumn(NORMAL_CF, QUALIFIER, VALUE);
+      table.put(put);
+    }
+    try (Table peerTable = utility2.getConnection().getTable(TABLE)) {
+      for (int i = 0; i < NB_RETRIES; i++) {
+        if (i == NB_RETRIES - 1) {
+          fail("Waited too much time for put replication");
+        }
+        Result result = peerTable.get(new Get(ROW).addColumn(NORMAL_CF, QUALIFIER));
+        if (result == null || result.isEmpty()) {
+          LOG.info("Row not available in peer cluster");
+          Thread.sleep(SLEEP_TIME);
+        } else {
+          assertArrayEquals(VALUE, result.getValue(NORMAL_CF, QUALIFIER));
+          break;
+        }
+      }
+    }
```

Comment:
HBaseTestingUtility has some nice `waitFor()` methods which simplify this kind of logic

Reply:
Thanks for tips, I've fixed it.